### DISCUSSION TITLE
How do you collage a bunch of images like a fanned deck of cards

### DISCUSSION THREAD

**Discussion by srini8080** [ImageMagick/ImageMagick#6992] [Digital Image Processing]
Centos 8.0
Version: ImageMagick 7.1.1-6 Q16-HDRI x86_64 21020 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
Delegates (built-in): bzlib djvu fontconfig freetype jbig jng jp2 jpeg lzma pangocairo png tiff webp x xml zip zlib
Compiler: gcc (8.5)

I have 20 images. I want to put them like a deck of cards fanned (collage). Like the image shown below.

They can be put randomly

![AcetoFive](https://github.com/ImageMagick/ImageMagick/assets/12046302/1427da4b-bbdd-4611-ac03-c72f589bc97a)

Thanks

**Comment by fmw42** [comment]
Post your images

**Reply by srini8080** [reply]
here they are

![27](https://github.com/ImageMagick/ImageMagick/assets/12046302/2f7fcdea-f3a2-4b5b-bdda-87e95f6ccfea)
![26](https://github.com/ImageMagick/ImageMagick/assets/12046302/e6fbb23a-25e5-42f4-9b34-53984196a28f)
![25](https://github.com/ImageMagick/ImageMagick/assets/12046302/fcd22528-4771-4d8a-8f60-170272a11e41)
![24](https://github.com/ImageMagick/ImageMagick/assets/12046302/f708d5b9-d769-43c4-a0ef-02bccb7d07a2)
![23](https://github.com/ImageMagick/ImageMagick/assets/12046302/ef790167-4a84-4205-aa09-b917b8c436b9)
![22](https://github.com/ImageMagick/ImageMagick/assets/12046302/e10431a1-da78-4795-8a62-1a7dcad33ec7)
![21](https://github.com/ImageMagick/ImageMagick/assets/12046302/69aead75-23e9-4b7a-9d9d-7a5d10fbc441)
![28](https://github.com/ImageMagick/ImageMagick/assets/12046302/3268401f-aac6-4bf6-8da2-8e82c7b4120d)
![29](https://github.com/ImageMagick/ImageMagick/assets/12046302/df78d59e-27bf-4a47-a524-15ae95752778)
![30](https://github.com/ImageMagick/ImageMagick/assets/12046302/124a9abb-b084-4b22-b662-ba2ab6f81085)
![22](https://github.com/ImageMagick/ImageMagick/assets/12046302/ddc1dfa5-7419-4aa8-8825-ff7272215299)

**Reply by fmw42** [reply]
I will try to work on it tomorrow.  Too late tonight.  But you can use distort SRT to rotate each about the bottom left corner in a FOR loop and then flatten all the results. (In IM 7, one could rotate them all in one command line if you have enough RAM, using inline fx computations for the rotation angle and the image number in the command line).

See

https://imagemagick.org/Usage/distorts/#srt
https://imagemagick.org/Usage/layers/#flatten

**Reply by srini8080** [reply]
Thanks Much, I will also explore.

Regards,

**Reply by fmw42** [reply]
Your images are not all the same size.  That may not look correct.  Do you want to resize to the same size?

**Comment by srini8080** [comment]
Sure. Say 1024x1024

⁣Get BlueMail for Android ​

**Comment by fmw42** [comment]
Here is something close

Unix syntax for IM 7.  Limited to 90 deg rotation.  Set for 60 deg.

```
cd
cd desktop/cards
i=0
num=10
(
for img in *.jpeg; do
magick "$img" \
-set option:W "%w" \
-background none \
-resize 1024x1024 \
-extent 1024x1024 \
-rotate 90 +repage \
-rotate "%[fx:60*$i/$((num-1))]" +repage \
-trim +repage \
-set page "-%[fx:W*sin((pi/180)*60*$i/$((num-1)))]+0" \
miff:-
i=$((i+1))
done
) | \
magick - \
-background none \
-layers merge +repage \
-rotate -120 \
card_deck.png

```

![card_deck](https://github.com/ImageMagick/ImageMagick/assets/13458218/34f9ac72-b2b1-4a25-9570-9f5e940fd2bd)

**Reply by srini8080** [reply]
worked the way I wanted. If we need to make this from 10 degrees to 170 degrees fan, where should I tweak?

**Reply by srini8080** [reply]
"-%[fx:W*sin((pi/180)*10*$i/$((num-1)))]+0" is what I wanted. Thanks much

**Comment by GeeMack** [comment]
With ImageMagick v7, given several images, you can generate something like a fan of cards using a "+distort" operation. Starting with 5 images like this one, cards-01.png, cards-02.png, etc...

![cards-02](https://github.com/ImageMagick/ImageMagick/assets/65475279/fbbd1387-4dbd-4341-82de-9f332fb2dacb)

... in a Windows CLI command like this...

    magick cards-*.png -background none -virtual-pixel none ^
       -gravity center -resize 320x320 -extent 320x320 +repage ^
       +distort SRT "%[fx:w*0.5],%[fx:h*1.1] 1 %[fx:t*60/(n-1)-30]" ^
       -layers merge result.png

... this is the result...

![result](https://github.com/ImageMagick/ImageMagick/assets/65475279/4ad05ad7-f7d6-440b-a0a0-504dca3f8b94)

The command reads in any practical number of images, then resizes and centers each in a canvas with the same dimensions.

The first pair of arguments to the "+distort" operation `%[fx:w*0.5],%[fx:h*1.1]` specifies the point for the pivot. In this command notice that point is below the bottom of the image. The "1" argument just means no change to the scale. The last argument `%[fx:t*60/(n-1)-30]` takes the number of input images and divides them evenly over the specified angle, in this case 60 degrees.

To run this as a Windows BAT script, double all the percent signs "%%".

To run the command on a *nix OS, change the continued-line carets "^" to a backslashes "\\".

**Reply by srini8080** [reply]
worked good with 30.

**Reply by fmw42** [reply]
Greg,  when I run your script converted to Unix, the images (cards) are upside down.

```
cd
cd desktop/cards
magick *.png -background none -virtual-pixel none \
-gravity center -resize 320x320 -extent 320x320 +repage \
+distort SRT "%[fx:w*0.5],%[fx:h*1.1] 1 %[fx:t*60/(n-1)-30]" \
-layers merge result.png

```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/1918f6ea-1698-4261-8d5e-62204ca8a38d)

**Reply by GeeMack** [reply]
That is odd. The command uses the plus form of "+distort", so it relies on IM to calculate how the rotations fit on the canvas. I don't see anything that might turn everything another 180 degrees.

**Reply by fmw42** [reply]
Do you get that result also, when using his images?

**Reply by GeeMack** [reply]
Using the exact command with IM v7.1.1-25 Q16-HDRI x64, and they didn't flip over. It's a mystery to me.

It looks like the image on the  bottom layer of your result, the top layer of mine here, didn't turn around in your example. It's also the only single PNG image. The rest are JPEGs. 

![result_test](https://github.com/ImageMagick/ImageMagick/assets/65475279/a51afc46-d884-4931-9c3f-22cdbc29074d)

**Reply by fmw42** [reply]
Resolved!

My mistake.  I was using magick * ... and forgot to clear out some test images in the folder. The images we were seeing in my above result were the test (resized images) at the bottom of my folder and so they covered the actual images that were provided.

So after clearing out the folder of other images, I get

```
cd
cd desktop/cards
magick * -background none -virtual-pixel none \
-gravity center -resize 320x320 -extent 320x320 +repage \
+distort SRT "%[fx:w*0.5],%[fx:h*1.1] 1 %[fx:t*60/(n-1)-30]" \
-layers merge result.png

```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/25d0c7b8-2e12-45ba-b963-e4edfa50be31)

**Comment by fmw42** [comment]
Here is a slightly improved version of my script

```
cd
cd desktop/cards
start_angle=-45
spread_angle=75   # 0 to 90
i=0
num=10
(
for img in *.jpeg; do
magick "$img" \
-background none \
-resize 1024x1024 \
-rotate 90 +repage \
-set option:H "%h" \
-rotate "%[fx:$spread_angle*$i/$((num-1))]" +repage \
-set page "-%[fx:-h/2+H*sin((pi/180)*$spread_angle*$i/$((num-1)))]+0" \
miff:-
i=$((i+1))
done
) | \
magick - \
-background none \
-layers merge +repage \
-rotate "%[fx:$start_angle-90]" \
card_deck.png

```

![card_deck](https://github.com/ImageMagick/ImageMagick/assets/13458218/5b802bfc-d0be-4c03-a5c0-59fc6571f6de)

**Comment by fmw42** [comment]
Here is a variation on Greg's command

```
cd
cd desktop/cards
start_angle=-45
spread_angle=240   # 0 to 360
image_dims=1024
magick *.jpeg \
-background none \
-resize 1024x1024 \
-rotate 90 +repage \
-virtual-pixel none \
-set option:distort:viewport %[fx:3*$image_dims]x%[fx:3*$image_dims]+0+0 \
-distort SRT "0,0 1 %[fx:$spread_angle*t/n] %[fx:1.5*$image_dims],%[fx:1.5*$image_dims]" \
-layers merge +repage \
-rotate "%[fx:$start_angle-90]" +repage \
card_deck2.png

```

![card_deck2](https://github.com/ImageMagick/ImageMagick/assets/13458218/2458b9ec-fe34-477b-947a-a7506aa6eaab)

**Reply by srini8080** [reply]
with rotate -90, this hit the nail on the head
